util/patricia_trie/src/lookup.rs

```diff
@@ -55,29 +55,32 @@ impl<'a, Q: Query> Lookup<'a, Q> {
 			// without incrementing the depth.
 			let mut node_data = &node_data[..];
 			loop {
-				match Node::decoded(node_data).expect("rlp read from db; qed") {
-					Node::Leaf(slice, value) => {
-						return Ok(match slice == key {
-							true => Some(self.query.decode(value)),
-							false => None,
-						})
-					}
-					Node::Extension(slice, item) => {
-						if key.starts_with(&slice) {
-							node_data = item;
-							key = key.mid(slice.len());
-						} else {
-							return Ok(None)
+				match Node::decoded(node_data) {
```

Comment:
I think this nested match could be avoided with a `?` here and change the `From` impl to `Box<TrieError>`?

Reply:
Good call, totally works! :)